### PR TITLE
enable cudagraph for varlen pretrain

### PR DESCRIPTION
**Pull request by liangel-02**
1. cuseqlens varies depending on number of docs --> pin to max_docs + 1 
2. max_q/max_k also vary --> pin to seq_len 
3. we call nonzero and the max().item() sync --> use cumsum instead

Screenshot 2026-08-14 at 6 21 08 PM

### CODE REVIEW

File: torchtitan/models/common/attention.py [hunk 1]
```diff
@@ -121,6 +121,12 @@ class Config(Module.Config):
               - (W, 0): Sliding window causal - attend to at most W previous tokens.
         """
 
+        max_num_documents: int | None = None
```
Comment (fegin) [inline torchtitan/models/common/attention.py]: noob question, what does the "kernel cost" mean here? Is it memory or computation?

Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: how are users supposed to set this field?

File: torchtitan/models/common/attention.py [hunk 2]
```diff
@@ -642,62 +649,101 @@ def create_varlen_metadata_for_document(
...
-    cu_seqlens_list, all_seq_lengths = [], []
-    offset = 0
+    packed_cu_seqlens_host = None
+
+    if max_num_documents is not None:
+        if include_host_offsets:
+            raise ValueError(
+                "include_host_offsets is incompatible with max_num_documents: "
+                "host offsets are data dependent."
+            )
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: Does it mean we couldn't support GDN with cudagraph? What's our plan?

File: torchtitan/models/common/attention.py [hunk 3]
```diff
@@ -642,62 +649,101 @@ def create_varlen_metadata_for_document(
...
+        if max_num_documents < batch_size:
+            raise ValueError(
+                f"max_num_documents ({max_num_documents}) must be at least the "
+                f"batch size ({batch_size})."
+            )
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: batch size is killed on main

File: torchtitan/trainer.py [hunk 4]
```diff
@@ -187,7 +187,23 @@ def _validate_cuda_graphs(self) -> None:
                     "Set --training.disable_cuda_graphs."
                 )
 
-            if self.parallelism.expert_parallel_degree == 1 or self.model_spec is None:
+            if self.model_spec is None:
+                return
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: assert not None

File: torchtitan/models/common/attention.py (hunk 1 above)
Comment (tianyu-l) [inline torchtitan/models/common/attention.py]: discussed with @drisspg 
- let's make this a config of dataloader, to cap the number of documents -- if the number of documents on a single local batch reaches this number, we can do padding (need to be careful with the interaction of padding and varlen? e.g. what if the length of padding is longer than `max_context_length`, etc.)
- pass this field from dataloder config to `post_dataloader_process` or after https://github.com/pytorch/torchtitan/pull/4116 `model.preprocess_inputs` -> then to `create_varlen_metadata_for_document`

File: torchtitan/trainer.py [hunk 5]
```diff
@@ -172,7 +172,25 @@ def _validate_cuda_graphs(self) -> None:
                     "Set --training.disable_cuda_graphs."
                 )
 
-            if self.parallelism.expert_parallel_degree == 1 or self.model_spec is None:
+            if self.model_spec is None:
+                assert not check_varlen_metadata
+                return
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: model_spec should never be None

File: torchtitan/trainer.py [hunk 6]
```diff
@@ -172,7 +172,25 @@ def _validate_cuda_graphs(self) -> None:
...
+
+            if check_varlen_metadata:
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: don't need this flag?

File: torchtitan/components/data/collators.py [hunk 7]
```diff
@@ -73,7 +75,58 @@ def __call__(self, rows: Sequence[TextSequence]) -> TrainerBatch:
...
+                max_context_length=self._max_context_length,
+            )
+
         return {
             "input": input_ids,
             "positions": positions,
         }, labels
+
+
+def _cap_document_count(
```
Comment (tianyu-l) [inline torchtitan/components/data/collators.py]: We should NOT drop any document.

In particular, I don't think this capping should happen in collator -- instead, it should happen when we construct each row.

File: torchtitan/trainer.py [hunk 8]
```diff
@@ -688,12 +706,25 @@ def forward_backward_step(
...
+            }
+            max_num_documents = getattr(
+                getattr(self.config, "dataloader", None),
+                "max_num_documents",
+                None,
+            )
+            if (
+                max_num_documents is not None
+                and not self.config.training.disable_cuda_graphs
+            ):
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: when cudagraph is disabled, is it not helpful to send in `max_num_documents`?

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/components/data/collators.py (hunk 7 above)
Comment (tianyu-l) [reply torchtitan/components/data/collators.py]: also need to modify RL's batcher, maybe on top of https://github.com/pytorch/torchtitan/pull/4326

File: torchtitan/experiments/rl/components/batcher.py [hunk 9]
```diff
@@ -92,6 +98,23 @@ def __init__(
...
+                "max_num_documents must be at least the number of rows in one "
+                f"local microbatch ({self.local_batch_size}), got "
+                f"{self.max_num_documents}"
+            )
+        # Rows are distributed round-robin into local microbatches. Giving every
+        # row the same share guarantees the local bound for any grid shape.
+        self._max_num_documents_per_row = (
+            self.max_num_documents // self.local_batch_size
+            if self.max_num_documents is not None
+            else None
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/components/batcher.py]: please refactor this when you remove the `local_batch_size` concept

File: torchtitan/experiments/rl/types.py [hunk 10]
```diff
@@ -130,6 +130,8 @@ class TrainingMicrobatch:
     generator_logprobs: torch.Tensor  # [T]
     loss_mask: torch.Tensor  # [T]
     advantages: torch.Tensor  # [T]
+    max_num_documents: int | None = None
+    """Fixed document-metadata capacity for this local microbatch."""
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/types.py]: I think having it in batcher config would suffice. See comment in rl trainer

File: torchtitan/experiments/rl/actors/trainer.py [hunk 11]
```diff
@@ -395,7 +395,10 @@ async def forward_backward(
         generator_logprobs = local_batch.generator_logprobs.to(device)
         advantages = local_batch.advantages.to(device)
 
-        attention_masks = model.get_attention_masks(positions)
+        attention_masks = model.get_attention_masks(
+            positions,
+            max_num_documents=local_batch.max_num_documents,
+        )
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py:406]: Right now I think we should only have one cudagraph for trainer. Let's send that from batcher config to the constructor of RL trainer.

File: torchtitan/trainer.py [hunk 12]
```diff
@@ -739,12 +759,22 @@ def pp_forward_backward_step(
...
             with sl.log_trace_span("preprocess_inputs"):
+                preprocess_kwargs = {
+                    "parallel_dims": self.parallel_dims,
+                    "parallelism": self.config.parallelism,
+                }
+                max_num_documents = getattr(
+                    getattr(self.config, "dataloader", None),
+                    "max_num_documents",
+                    None,
+                )
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: how about just inline this in `preprocess_inputs`

File: torchtitan/trainer.py [hunk 13]
```diff
@@ -688,12 +698,22 @@ def forward_backward_step(
...
         with sl.log_trace_span("preprocess_inputs"):
+            preprocess_kwargs = {
+                "parallel_dims": self.parallel_dims,
+                "parallelism": self.config.parallelism,
+            }
+            max_num_documents = getattr(
+                getattr(self.config, "dataloader", None),
+                "max_num_documents",
+                None,
+            )
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: similar nit

File: torchtitan/trainer.py [hunk 14]
```diff
@@ -172,7 +175,18 @@ def _validate_cuda_graphs(self) -> None:
                     "Set --training.disable_cuda_graphs."
                 )
 
-            if self.parallelism.expert_parallel_degree == 1 or self.model_spec is None:
+            max_num_documents = getattr(self.dataloader, "max_num_documents", None)
+            for fqn, _, _, _ in self.model_spec.model.traverse(VarlenAttention.Config):
+                if max_num_documents is None:
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: I think we should reverse logic -- if `max_num_documents is None`, then run the traverse to search for varlen and raise

File: torchtitan/components/data/packing.py
Comment (tianyu-l) [inline torchtitan/components/data/packing.py:1]: looks complicated

**Comment by tianyu-l** [comment]
@claude review this PR.

In particular, I found the logic in torchtitan/components/data/packing.py complicated. Could it be simplified?

Overall, I would like to make sure
- Documents longer than `max_context_length` is properly treated, either filtered or truncated. What is being done today and what's your recommendation?
- If a single row has `max_num_documents` documents to pack and they can all fit, we should pad the rest. What is the current behavior? We should make sure we don't "lose" any documents (unless it's too long, see above point). Also, does it work nicely with changes in torchtitan/models/common/attention.py? I think we should output padding mask from dataloader and RL batcher, so that attention metadata building (and later MoE router logic regarding load balancing) can be made clear.

File: torchtitan/components/data/collators.py [hunk 15]
```diff
@@ -64,16 +65,32 @@ def __call__(self, rows: Sequence[TextSequence]) -> TrainerBatch:
...
             ]
         )
+        padding_mask = torch.cat(
+            [
+                torch.zeros(len(row.input_ids), dtype=torch.bool)
+                if row.padding_mask is None
+                else torch.as_tensor(row.padding_mask, dtype=torch.bool)
+                for row in rows
+            ]
+        )
```
Comment (tianyu-l) [inline torchtitan/components/data/collators.py:75]: This seems to suggest that `positions` and `padding_mask` can be handled either before collator or by collator, correct?

File: torchtitan/experiments/rl/components/batcher.py [hunk 16]
```diff
@@ -309,6 +329,53 @@ def _build_microbatch_grid(
             grid.append(ranks)
         return grid
 
+    def _build_document_limited_microbatch_grid(
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/components/batcher.py:332]: didn't check the internal of this function, but I believe things could be made simpler when we fold the batch dim in each microbatch @wwwjn

File: torchtitan/models/common/attention.py [hunk 17]
```diff
@@ -601,51 +604,109 @@ def create_varlen_metadata_for_document(
...
+        padding_doc_starts = (positions == 0) & padding_mask
+        is_doc_start = real_doc_starts | padding_doc_starts
+
+    if max_num_documents is not None:
+        if max_context_length is None:
+            raise ValueError(
+                "max_context_length is required when max_num_documents is set"
+            )
+
+        max_num_padding_segments = (
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py:644]: I didn't understand what this count is for. Here's my guess but it could be off.

There are generally two types of padding
1. per document padding, e.g. needed by FlexAttention to achieve batch invariance
2. per sequence / row padding, at the end of a (row of a) microbatch

I originally thought varlen meta data only needs to know about 2, but it seems here you are accounting for 1 as well? What's the tradeoff of varlen metadata to consider 1 vs. not? Naively it sounds to me that
- knowing 1 would help varlen to compute less (no attention on the per document padding)
- knowing 1 would require varlen to allocate more resource to take care of padding

Would you argue knowing 1 is strictly better?

**Review by tianyu-l — APPROVED** [review]
overall looks great. Having some remaining questions but don't want to block merging

**Comment by tianyu-l** [comment]
@claude 
For this new version, briefly state what happens for the two packing strategies. What I think we should achieve in both
- no dropping of real data, unless it surpasses the `min(tokens_per_batch, max_context_length)` limit set by model
- for the padding one, we start the data that couldn't fit on a new row
- for the chunking one, we fit whatever we could on current row and start the rest on next row (but continue the positions counting)

Also take a look at my comment https://github.com/pytorch/torchtitan/pull/4156#discussion_r3887847182 and provide insight.

**Review by tianyu-l — APPROVED** [review]
